### ISSUE TITLE
string.c_str doesn't return cstring of Node

### ISSUE THREAD

**Issue by Zerophase** [jbeder/yaml-cpp#491]
I have a project that needs to pass cstrings to other libraries, with `c_str()`.

This code doesn't grab the string:
```c++
const char *cstring = config["Variable"].as<std::string>().c_str();
```
Even if the variable is saved to a string a cstring still cannot be grabbed:
```c++
string str = config["Variable"].as<std::string>();
std::cout << str.c_str() << std::endl // prints nothing
```

**Comment by jbeder** [comment]
Is the std:string itself nonempty?

**Comment by Zerophase** [comment]
Yeah it's not empty, if I don't call c_str. It seems to work sometimes, and not others depending on how the string is accessed.

**Comment by keith-bennett-gbg** [comment]
`const char *cstring = ...` will invoke undefined behavior. You are grabbing a pointer to a temporary. The temporary is deallocated after the next full sequence point (semicolon), leaving `cstring` dangling.

**Comment by jbeder** [comment]
@keith-bennett-gbg is correct; your answer to my question illustrates it. Just keep the std::string around (your second example probably works fine).

**Comment by Zerophase** [comment]
@keith-bennett-gbg `config["Variable"].Scalar().c_str()` has the same effect. I'm just trying to avoid having to include the `string` header if there is anyway to get away without it. For now, I'm just saving to a `string` and calling `c_str()` when I need to pass in a cstring.

**Comment by keith-bennett-gbg** [comment]
> config["Variable"].Scalar().c_str() has the same effect.

I'll respond to this in a moment.

> I'm just trying to avoid having to include the string header if there is anyway to get away without it.

I'd be willing to bet that it's already being `#include`-ed from one of the yaml-cpp headers. For example, if you use `#include <yaml-cpp/yaml.h>`, then

1. `yaml.h` -> [`#include <yaml-cpp/node/impl.h>`](https://github.com/jbeder/yaml-cpp/blob/11607eb5bf1258641d80f7051e7cf09e317b4746/include/yaml-cpp/yaml.h#L17)
2. `impl.h` -> [`#include <string>`](https://github.com/jbeder/yaml-cpp/blob/11607eb5bf1258641d80f7051e7cf09e317b4746/include/yaml-cpp/node/impl.h#L15)

> For now, I'm just saving to a string and calling c_str() when I need to pass in a cstring.

The tl;dr is that this is the right thing to do if you suspect lifetime issues are causing problems.


Okay, so in response to your first point:
So, admittedly the code is not IMO easy to reason about. So I'm not going to say I'm _not_ wrong. But I didn't write the library so I definitely _could_ be wrong. Here's how I'm thinking that I'm correct:

1. [`config["Variable"]` returns a `Node` or a `const Node`](https://github.com/jbeder/yaml-cpp/blob/11607eb5bf1258641d80f7051e7cf09e317b4746/include/yaml-cpp/node/node.h#L108-L109)
2. [`Node::Scalar()` returns a `const std::string &`](https://github.com/jbeder/yaml-cpp/blob/11607eb5bf1258641d80f7051e7cf09e317b4746/include/yaml-cpp/node/node.h#L69).
3. _Actually_, [`Node::Scalar()`](https://github.com/jbeder/yaml-cpp/blob/11607eb5bf1258641d80f7051e7cf09e317b4746/include/yaml-cpp/node/impl.h#L156-L160) throws if the `Node` is not valid. Otherwise, it checks if it has an underlying `detail::node`. If so, it returns `detail::node::scalar()`. If it does not, then it returns [`detail::node_data::empty_scalar`, a `static` variable](https://github.com/jbeder/yaml-cpp/blob/11607eb5bf1258641d80f7051e7cf09e317b4746/include/yaml-cpp/node/detail/node_data.h#L84). I'm going to _assume_ your case isn't returning the `static`-ally allocated string.
4. [`detail::node::scalar()` also returns `const std::string &`](https://github.com/jbeder/yaml-cpp/blob/11607eb5bf1258641d80f7051e7cf09e317b4746/include/yaml-cpp/node/detail/node.h#L32) and has the rules from point 2 above applied; this time it's returned from `detail::node_ref::scalar()`.
5. [`detail::node_ref::scalar()` _also_ returns `const std::string &`](https://github.com/jbeder/yaml-cpp/blob/11607eb5bf1258641d80f7051e7cf09e317b4746/include/yaml-cpp/node/detail/node_ref.h#L26), this time from `std::shared_ptr<node_data>::scalar()`. Note that [`std::shared_ptr<node_data>` gets typedefed to `shared_node_data`](https://github.com/jbeder/yaml-cpp/blob/11607eb5bf1258641d80f7051e7cf09e317b4746/include/yaml-cpp/node/ptr.h#L23). Wow we're getting pretty deep into the guts here.
6. [`detail::node_data` returns `m_scalar`](https://github.com/jbeder/yaml-cpp/blob/11607eb5bf1258641d80f7051e7cf09e317b4746/include/yaml-cpp/node/detail/node_data.h#L49), which is the final and actual location of the [`std::string`](https://github.com/jbeder/yaml-cpp/blob/11607eb5bf1258641d80f7051e7cf09e317b4746/include/yaml-cpp/node/detail/node_data.h#L108) _reference_ that gets returned.

Okay, so now we know _where_ it's at, we can _try_ to reason about its lifetime. This is where it gets tricky, particularly with respect to `const`ness. I'd love to learn the design decision tree for this, but suffice to say that I disagree with it. Anyway...

So since it's a member variable of `detail::node_data`, what's the lifetime of that? Given that it's accessed through a `shared_ptr`, I suspect its lifetime is controlled by your `config` variable. That is, I believe that as long as `config` is alive (and you haven't deleted the specific subnode named `"Variable"`), so should the `shared_ptr` for `detail::node_data`. Let's check this out.

1. `Node::operator[]` returns a new (possibly-`const`) `Node`. [The implementation](https://github.com/jbeder/yaml-cpp/blob/11607eb5bf1258641d80f7051e7cf09e317b4746/include/yaml-cpp/node/impl.h#L371-L392) seems to look up the `detail::node` and then construct a new Node with that. In the `const` version, it forces local `mutable m_pNode` to be const, then retrieves `detail::node::get()`, passing `m_pMemory`. The non-`const` version does the same thing without forcing `m_pNode` to `const`.
2. [`detail::node::get()` also returns a different value based on `const`ness](https://github.com/jbeder/yaml-cpp/blob/11607eb5bf1258641d80f7051e7cf09e317b4746/include/yaml-cpp/node/detail/node.h#L119-L132). Again, it doesn't really make much sense to me and this is where I get lost: we end up going again through several levels of indirection (ugh, performance is out of the window). Frankly, I don't care enough to go digging that deep into the lifetime of each object here.
3. The returned value gets passed to [`Node`'s private constructor, `Node::Node(detail::node& node, detail::shared_memory_holder)`](https://github.com/jbeder/yaml-cpp/blob/11607eb5bf1258641d80f7051e7cf09e317b4746/include/yaml-cpp/node/node.h#L119). The implementation [just hands those off to internal member variables](https://github.com/jbeder/yaml-cpp/blob/11607eb5bf1258641d80f7051e7cf09e317b4746/include/yaml-cpp/node/impl.h#L47-L48).

The final `Node` private constructor accepts a `shared_memory_holder` (`shared_ptr`) and a `detail::node` reference. In both cases, the `shared_memory_holder` passed to the returned (newly-constructed) Node is the _same_ instance as `config.m_pMemory`. It _appears_ to me that the `detail::node` reference has the same lifetime as well, if perhaps retrieved in different ways.

**tl;dr**:
All of this complexity and deep thinking can be skipped by copying your needed `std::string` to somewhere you control and returning the `.c_str()` from that.
